### PR TITLE
feat: upgrade BQ managed `udf` to preview

### PR DESCRIPTION
**Pull request by shobsi**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: bigframes/functions/_function_client.py [hunk 1]
```diff
@@ -193,11 +196,20 @@ def provision_bq_managed_function(
...
     ):
         """Create a BigQuery managed function."""
-        import cloudpickle
 
-        pickled = cloudpickle.dumps(func)
+        if capture_references:
+            # Check the Python version.
+            python_version = _utils.get_python_version()
+            if python_version != _MANAGED_FUNC_PYTHON_VERSION:
+                raise bf_formatting.create_exception_with_feedback_link(
```
Comment (jialuoo) [inline bigframes/functions/_function_client.py:208]: In this error message, we should add more details about "capture_references=True cases has limitation on versions ...".

File: bigframes/functions/_function_client.py [hunk 2]
```diff
@@ -193,11 +196,20 @@ def provision_bq_managed_function(
         name,
         packages,
         is_row_processor,
+        *,
+        capture_references=False,
```
Comment (jialuoo) [inline bigframes/functions/_function_client.py:200]: nit: add comments about why we need capture_references.

File: bigframes/functions/_function_session.py [hunk 3]
```diff
@@ -809,17 +806,8 @@ def udf(
...
 
-        # Check the Python version.
-        python_version = _utils.get_python_version()
-        if python_version not in _MANAGED_FUNC_PYTHON_VERSIONS:
-            raise bf_formatting.create_exception_with_feedback_link(
-                RuntimeError,
-                f"Python version {python_version} is not supported yet for "
-                "BigFrames managed function.",
-            )
+        warnings.warn("udf is in preview.", category=bfe.PreviewWarning)
```
Comment (jialuoo) [inline bigframes/functions/_function_session.py:810]: This is a user-facing warning. To maintain consistency, I suggest we replace 'udf' with 'managed function' (I feel I used latter one in some other user-facing messages).

I'm flexible on which term we use, but it should be consistent.

File: tests/system/large/functions/test_managed_function.py [hunk 4]
```diff
@@ -121,21 +120,12 @@ def stringify(x):
         cleanup_function_assets(stringify, bigquery_client, ignore_failures=False)
 
 
-@pytest.mark.parametrize(
```
Comment (jialuoo) [inline tests/system/large/functions/test_managed_function.py:130]: Why do we remove this?

File: bigframes/functions/_function_client.py [hunk 5]
```diff
@@ -193,11 +196,20 @@ def provision_bq_managed_function(
...
         """Create a BigQuery managed function."""
-        import cloudpickle
 
-        pickled = cloudpickle.dumps(func)
+        if capture_references:
+            # Check the Python version.
+            python_version = _utils.get_python_version()
+            if python_version != _MANAGED_FUNC_PYTHON_VERSION:
+                raise bf_formatting.create_exception_with_feedback_link(
+                    RuntimeError,
```
Comment (jialuoo) [inline bigframes/functions/_function_client.py:209]: Should it be NotImplementedError?

File: tests/system/large/functions/test_managed_function.py
Comment (jialuoo) [inline tests/system/large/functions/test_managed_function.py:1]: We should add a test for 'none self contained udf'.

File: bigframes/functions/_function_client.py (hunk 2 above)
Comment (shobsi) [reply bigframes/functions/_function_client.py:200]: It's there in the if-else clause down below (line#204)

File: bigframes/functions/_function_session.py (hunk 3 above)
Comment (shobsi) [reply bigframes/functions/_function_session.py:810]: I changed "managed function" -> "udf" in couple of messages, PTAL. I agree it should be consistent. Let's use "udf" to reflect the BigFrames API, as that is what it is named, and consistently maintain that it is built on top of "BigQuery managed function".

File: tests/system/large/functions/test_managed_function.py (hunk 4 above)
Comment (shobsi) [reply tests/system/large/functions/test_managed_function.py:130]: Because now we don't support the outside references, parameterize comes as outside reference and resolved dynamically, which was handled well with cloudpickle but not anymore. Once we expose outside reference code paths to user again we will be able to use parametrize again.

File: bigframes/functions/_function_client.py (hunk 5 above)
Comment (shobsi) [reply bigframes/functions/_function_client.py:209]: Updated, thanks!

File: tests/system/large/functions/test_managed_function.py
Comment (shobsi) [reply tests/system/large/functions/test_managed_function.py:1]: This functionality is not exposed in the public `udf` API (even though we kept the code path internally to expose later b/406283812). It will require extra consideration to test restricted to the supported python version, let's defer to that work.

File: bigframes/functions/_function_client.py (hunk 1 above)
Comment (shobsi) [reply bigframes/functions/_function_client.py]: Rephrased the error message, PTAL, thanks!

**Review by jialuoo — APPROVED** [review]
Nice job!

**Comment by shobsi** [comment]
Even though the managed udfs tests are disabled currently, I ran them locally successfully.
